File: backend/py_template/devdonalds.py
from dataclasses import dataclass
from typing import List, Dict, Union
from flask import Flask, request, jsonify
from collections import deque, defaultdict
import re, pprint

# ==== Type Definitions, feel free to add or modify ===========================
@dataclass
class CookbookEntry:
    name: str

@dataclass
class RequiredItem():
    name: str
    quantity: int

@dataclass
class Recipe(CookbookEntry):
    required_items: List[RequiredItem]

@dataclass
class Ingredient(CookbookEntry):
    cook_time: int


# =============================================================================
# ==== HTTP Endpoint Stubs ====================================================
# =============================================================================
app = Flask(__name__)

# Store your recipes here!
cookbook = {}

# Task 1 helper (don't touch)
@app.route("/parse", methods=['POST'])
def parse():
    data = request.get_json()
    recipe_name = data.get('input', '')
    parsed_name = parse_handwriting(recipe_name)
    if parsed_name is None:
        return 'Invalid recipe name', 400
    return jsonify({'msg': parsed_name}), 200

# [TASK 1] ====================================================================
# Takes in a recipeName and returns it in a form that 
def parse_handwriting(recipeName: str) -> Union[str | None]:
    # 0. No alphabetical characters should be considered invalid
    if not re.search(r'[a-zA-Z]', recipeName):
        return None

    # 1. Replace - & _ with whitespace
    recipeName = recipeName.replace('-', ' ')
    recipeName = recipeName.replace('_', ' ')

    # 2. Non-alphabetic characters removed
    recipeName = re.sub(r'[^a-zA-Z ]', '', recipeName)
 
    # 3. Make all lowercase
    recipeName = recipeName.lower()

    # 4. Multiple whitespaces should be truncated to one
    recipeName = re.sub(r'\s+', ' ', recipeName)

    # 5. First character capitalised
    recipeName = recipeName.title()

    return recipeName

# [TASK 2] ====================================================================
# Endpoint that adds a CookbookEntry to your magical cookbook
@app.route('/entry', methods=['POST'])
def create_entry():
    data = request.get_json()

    if 'name' not in data or 'type' not in data:
        return jsonify({'msg': 'error found'}), 400

    # Maybe data['name'] can be invalid
    entry_name = parse_handwriting(data['name'])
    if entry_name is None:
        return jsonify({'msg': 'error found'}), 400

    # Check for duplicate entry name
    if entry_name in cookbook:
        return jsonify({'msg': 'error found'}), 400

    entry_type = data['type']

    # Error handling for recipe
    if entry_type == 'recipe':
        # Extra error handling (type safety)
        if 'requiredItems' not in data or not isinstance(data['requiredItems'], list):
            return jsonify({'msg': 'error found'}), 400

        success, recipe_obj = parse_recipe(entry_name, data['requiredItems'])

        if success:
            # Add to cookbook
            cookbook[entry_name] = recipe_obj
            return jsonify({}), 200
        return jsonify({'msg': 'error found'}), 400

    # Error handling for ingredient
    elif entry_type == 'ingredient':
        if 'cookTime' not in data or not isinstance(data['cookTime'], int):
            return jsonify({'msg': 'error found'}), 400

        success, ingredient_obj = parse_ingredient(entry_name, data['cookTime'])

        if success:
            # Add to cookbook
            cookbook[entry_name] = ingredient_obj            
            return jsonify({}), 200
        return jsonify({'msg': 'error found'}), 400
    else:
	    return jsonify({'msg': 'error found'}), 400

def parse_recipe(name, required_items):
    # Checking for duplicates with set()
    required_items_set = set()
    parsed_items = []

    for item in required_items:
        # Error handling, just in case recipe's JSON format is invalid
        if 'name' not in item or 'quantity' not in item:
            return False, None

        if not isinstance(item['name'], str) or not isinstance(item['quantity'], int) or item['quantity'] <= 0:
            return False, None
        
        # Just in case name is invalid format        
        parsed_name = parse_handwriting(item['name'])
        if parsed_name in required_items_set:
            return False, None

        required_items_set.add(parsed_name)
        parsed_items.append(RequiredItem(name=parsed_name, quantity=item['quantity']))

    return True, Recipe(name=name, required_items=parsed_items)

def parse_ingredient(name, cook_time):
    if cook_time < 0:
        return False, None
    return True, Ingredient(name=name, cook_time=cook_time)


# [TASK 3] ====================================================================
# Endpoint that returns a summary of a recipe that corresponds to a query name
@app.route('/summary', methods=['GET'])
def summary():
    # Plan:
    # 1. Basic Error Handling
    #    Traverse through the current cookbook 1 (Naive search)
    #   - Searched name is not found
    #   - Searched name is an ingredient
    # 2. Additional error handling
    #    Traverse through the current cookbook 2 (BFS)
    #   - Recipe's requiredItems has not been satisfied
    #   - All leaf nodes are ingredients (additional field check)
    #       - If not leaf node, we cannot calculate quantities
    # 3. During traversal 2, for each leaf node,
    #    make the new Recipe structure, with the whole cookTime and ingredients.

    name = request.args.get('name', str)
    
    valid, _ = traverse(name, 1)
    
    if not valid:
        return jsonify({'msg': 'error found'}), 400
    
    valid, recipe = traverse(name, 2)
    
    if not valid:
        return jsonify({'msg': 'error found'}), 400

    return jsonify(recipe), 200

# Returns error (bool), recipe (raw json)
def traverse(target_name: str, option: int):
    # Option 1 for traversal 1, Option 2 for traversal 2, Option 3 for debugging and printing everything
    if option != 1 and option != 2 and option != 3:
        return False

    if option == 1:
        # Target is not in cookbook
        if target_name not in cookbook:
            return False, None

        entry = cookbook[target_name]
        # Target is an ingredient
        if isinstance(entry, Ingredient):
            return False, None

        # Target is a valid recipe
        return True, None

    elif option == 2:
        entry = cookbook[target_name]
        
        # Maps ingredient name to total quantity, missing key of defaultdict(int) returns 0
        ingredient_count = defaultdict(int)
        
        # Conduct a BFS
        queue = deque()
        queue.append((entry, 1))
        
        while queue:
            current, multiplier = queue.popleft()
            
            if isinstance(current, Ingredient):
                ingredient_count[current.name] += multiplier
            elif isinstance(current, Recipe):
                for required_item in current.required_items:
                    # Error if cookbook doesn't contain that required item
                    if required_item.name not in cookbook:
                        return False, None
                    
                    child = cookbook[required_item.name]
                    queue.append((child, multiplier * required_item.quantity))
            else:
                return False, None
        
        # Calculate total cooking time
        total_cooktime = 0
        ingredients_list = []
        
        # Build ingredients list
        for ingredient, quantity in ingredient_count.items():
            # We don't want recipes left, only need pure ingredients to calculate total cooking time
            if ingredient not in cookbook or not isinstance(cookbook[ingredient], Ingredient):
                return False, None
            
            total_cooktime += quantity * cookbook[ingredient].cook_time
            ingredients_list.append({
                "name": ingredient,
                "quantity": quantity
            })
        
        formatted_recipe = {
            "name": target_name,
            "cookTime": total_cooktime,
            "ingredients": ingredients_list
        }
        return True, formatted_recipe

    elif option == 3:
        pprint.pprint(cookbook)
        return True, None

# =============================================================================
# ==== DO NOT TOUCH ===========================================================
# =============================================================================

if __name__ == '__main__':
    app.run(debug=True, port=8080)

File: backend/ts_template/devdonalds.ts
import express, { Request, Response } from "express";

// ==== Type Definitions, feel free to add or modify ==========================
interface cookbookEntry {
  name: string;
  type: string;
}

interface requiredItem {
  name: string;
  quantity: number;
}

interface recipe extends cookbookEntry {
  requiredItems: requiredItem[];
}

interface ingredient extends cookbookEntry {
  cookTime: number;
}

// =============================================================================
// ==== HTTP Endpoint Stubs ====================================================
// =============================================================================
const app = express();
app.use(express.json());

// Store your recipes here!
const cookbook: any = {};

// Task 1 helper (don't touch)
app.post("/parse", (req: Request, res: Response) => {
  const { input } = req.body;

  const parsed_string = parse_handwriting(input)
  if (parsed_string == null) {
    res.status(400).send("this string is cooked");
    return;
  } 
  res.json({ msg: parsed_string });
  return;
  
});

// [TASK 1] ====================================================================
// Takes in a recipeName and returns it in a form that 
const parse_handwriting = (recipeName: string): string | null => {
  if (!/[a-zA-Z]/.test(recipeName)) {
    return null;
  }
  // 1. Replace - & _ with whitespace
  recipeName = recipeName.replace(/[-_]/g, " ");
  
  // 2. Non-alphabetic characters removed
  recipeName = recipeName.replace(/[^a-zA-Z ]/g, "");
  
  // 3. Make all lowercase
  recipeName = recipeName.toLowerCase();
  
  // 4. Multiple whitespaces should be truncated to one
  recipeName = recipeName.replace(/\s+/g, " ").trim();
  
  // 5. First character capitalised
  recipeName = recipeName
    .split(" ")
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
    .join(" ");

  return recipeName;
};
// [TASK 2] ====================================================================
// Endpoint that adds a CookbookEntry to your magical cookbook
app.post("/entry", (req: Request, res: Response) => {
  const data = req.body;
  
  // Basic error handling
  if (!data.name || !data.type) {
    res.status(400).json({ msg: "error found" });
    return;
  }

  // Maybe data['name'] can be invalid
  const entryName = parse_handwriting(data.name);
  if (entryName === null) {
    res.status(400).json({ msg: "error found" });
    return;
  }

  // Check for duplicate entry name
  if (cookbook[entryName]) {
    res.status(400).json({ msg: "error found" });
    return;
  }

  // Error handling for recipe
  if (data.type === "recipe") {
    if (!data.requiredItems || !Array.isArray(data.requiredItems)) {
      res.status(400).json({ msg: "error found" });
      return;
    }
    const [success, recipeObj] = parseRecipe(entryName, data.requiredItems);
    if (success && recipeObj) {
      cookbook[entryName] = recipeObj;
      res.status(200).json({});
      return;
    }
    res.status(400).json({ msg: "error found" });
    return;
  }

  // Error handling for ingredient
  else if (data.type === "ingredient") {
    if (data.cookTime === undefined || typeof data.cookTime !== "number") {
      res.status(400).json({ msg: "error found" });
      return;
    }
    const [success, ingredientObj] = parseIngredient(entryName, data.cookTime);
    if (success && ingredientObj) {
      cookbook[entryName] = ingredientObj;
      res.status(200).json({});
      return;
    }
    res.status(400).json({ msg: "error found" });
    return;
  } 

  // Everything else should be an error
  else {
    res.status(400).json({ msg: "error found" });
    return;
  }
});

const parseRecipe = (
  name: string,
  requiredItems: any[]
): [boolean, recipe | null] => {
  const requiredItemsSet = new Set<string>();
  const parsedItems: requiredItem[] = [];

  for (const item of requiredItems) {
    if (!item.hasOwnProperty("name") || !item.hasOwnProperty("quantity")) {
      return [false, null];
    }
    if (
      typeof item.name !== "string" ||
      typeof item.quantity !== "number" ||
      item.quantity <= 0
    ) {
      return [false, null];
    }
    const parsedName = parse_handwriting(item.name);
    if (parsedName === null) {
      return [false, null];
    }
    if (requiredItemsSet.has(parsedName)) {
      return [false, null];
    }
    requiredItemsSet.add(parsedName);
    parsedItems.push({ name: parsedName, quantity: item.quantity });
  }

  const recipeObj: recipe = {
    name: name,
    type: "recipe",
    requiredItems: parsedItems,
  };

  return [true, recipeObj];
};

const parseIngredient = (
  name: string,
  cookTime: number
): [boolean, ingredient | null] => {
  if (cookTime < 0) {
    return [false, null];
  }
  const ingredientObj: ingredient = {
    name: name,
    type: "ingredient",
    cookTime: cookTime,
  };
  return [true, ingredientObj];
};


// [TASK 3] ====================================================================
// Endpoint that returns a summary of a recipe that corresponds to a query name
app.get("/summary", (req: Request, res: Request) => {
  // The logic for this code is well-documented in devdonalds.py

  const nameQuery = req.query.name;
  if (!nameQuery || typeof nameQuery !== "string") {
    res.status(400).json({ msg: "error found" });
    return;
  }
  const name = nameQuery;

  const [valid1, _] = traverse(name, 1);
  if (!valid1) {
    res.status(400).json({ msg: "error found" });
    return;
  }
  const [valid2, recipe] = traverse(name, 2);
  if (!valid2 || recipe === null) {
    res.status(400).json({ msg: "error found" });
    return;
  }
  res.status(200).json(recipe);
});

function traverse(targetName: string, option: number): [boolean, any] {
  if (option !== 1 && option !== 2) {
    return [false, null];
  }

  // Option 1: Check that the target exists and is a recipe.
  if (option === 1) {
    if (!cookbook[targetName]) {
      return [false, null];
    }
    const entry = cookbook[targetName];
    if (entry.type === "ingredient") {
      return [false, null];
    }
    return [true, null];
  }
  
  // Option 2: Perform BFS to calculate total cookTime and ingredients.
  else if (option === 2) {
    const entry = cookbook[targetName];
    const ingredientCount: { [key: string]: number } = {};
    const queue: Array<[recipe | ingredient, number]> = [];
    queue.push([entry, 1]);

    while (queue.length > 0) {
      const [current, multiplier] = queue.shift()!;
      if (current.type === "ingredient") {
        ingredientCount[current.name] =
          (ingredientCount[current.name] || 0) + multiplier;
      } else if (current.type === "recipe") {
        for (const requiredItem of (current as recipe).requiredItems) {
          if (!cookbook[requiredItem.name]) {
            return [false, null];
          }
          const child = cookbook[requiredItem.name];
          queue.push([child, multiplier * requiredItem.quantity]);
        }
      } else {
        return [false, null];
      }
    }

    let totalCookTime = 0;
    const ingredientsList: Array<{ name: string; quantity: number }> = [];
    for (const ingredientName in ingredientCount) {
      const entry = cookbook[ingredientName];
      if (!entry || entry.type !== "ingredient") {
        return [false, null];
      }
      totalCookTime += ingredientCount[ingredientName] * (entry as ingredient).cookTime;
      ingredientsList.push({
        name: ingredientName,
        quantity: ingredientCount[ingredientName],
      });
    }

    const formattedRecipe = {
      name: targetName,
      cookTime: totalCookTime,
      ingredients: ingredientsList,
    };
    return [true, formattedRecipe];
  }
  return [false, null];
}

// =============================================================================
// ==== DO NOT TOUCH ===========================================================
// =============================================================================
const port = 8080;
app.listen(port, () => {
  console.log(`Running on: http://127.0.0.1:8080`);
});
